### PR TITLE
Add version.txt support

### PR DESCRIPTION
**Pull request by rjfd**
This PR moves the source of the version number from the spec file to the file `version.txt`, which is also installed by the RPM and Makefile installation.

The tarball generation implemented in the Makefile is also changed to use the new version number location. The tarball generation, besides creating the tarball it also creates a deepsea spec file with a version number that is composed by:
 * the string inside `version.txt` file (DeepSea release number)
* the offset (number of commits since the last commit that changed the `version.txt` file)
 * the commit hash

Signed-off-by: Ricardo Dias <rdias@suse.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -585,8 +587,17 @@ rpm: tarball test
...
 # Removing test dependency until resolved
 tarball:
-	VERSION=`awk '/^Version/ {print $$2}' deepsea.spec`; \
-	git archive --prefix deepsea-$$VERSION/ -o ~/rpmbuild/SOURCES/deepsea-$$VERSION.tar.bz2 HEAD
+	mkdir -p .tmp
+	git archive --prefix deepsea-tmp/ -o .tmp/deepsea-tmp.tar HEAD
+	HASH=`git rev-parse --short HEAD`; \
+	cd .tmp; tar -xf deepsea-tmp.tar; cd deepsea-tmp; \
+	VERSION=`cat version.txt`; \
+	echo "$$VERSION+git.$$HASH" > version.txt; \
```
Comment (smithfarm) [inline Makefile]: This is problematic because it does not guarantee that each successive build will have an ascending version number. You probably want to add a "git offset" component which can be obtained by parsing the output of `git describe`.

I thought `version.txt` was going to include just the "upstream" version number (i.e. "0.8" - shouldn't this be "0.8.0" though?) and the git offset + sha1 part would be appended to it in the packaging?

Comment (smithfarm) [reply Makefile]:
```
+	VERSION=`cat version.txt`; \
+	echo "$$VERSION+git.$$HASH" > version.txt; \
```

whoa, doesn't this overwrite `version.txt`? What happens if I run "make tarball" 3 times in a row?

Comment (smithfarm) [reply Makefile]: ah, nevermind, it's doing it in a temporary directory. So, ignore all the comments above, except for the "ascending version number" one :-)

Comment (smithfarm) [reply Makefile]: Note that RPM will refuse to upgrade version "0.8+git.81234567" to "0.8+git.12345678" even if the latter is newer (git SHA1s are essentially random).

Also, once you are parsing `git describe`, you might as well get the SHA1 from that, too.

Comment (rjfd) [reply Makefile]: My problem with the getting the offset is, which base should I compute the offset from?
The last defined tag?

Comment (smithfarm) [reply Makefile]: Indeed, using `git describe` assumes that the "upstream" version number will be incremented whenever a new tag is added - otherwise the addition of any new tag causes the offset to be reset to zero. Very good point. If you want something foolproof, use "seconds since epoch" instead of the offset.

I was/am just in the process of switching the ceph OBS build from "seconds since epoch" to "offset" but now I see we're going to have the same problem there.

Comment (rjfd) [reply Makefile]: I pushed an offset computation that computes the offset since the last commit that changed the `version.txt` file. That way it is not dependent on a git tag but rather on the version bump commit

File: Makefile [hunk 2]
```diff
@@ -585,8 +587,18 @@ rpm: tarball test
...
 # Removing test dependency until resolved
 tarball:
-	VERSION=`awk '/^Version/ {print $$2}' deepsea.spec`; \
-	git archive --prefix deepsea-$$VERSION/ -o ~/rpmbuild/SOURCES/deepsea-$$VERSION.tar.bz2 HEAD
+	mkdir -p .tmp
+	git archive --prefix deepsea-tmp/ -o .tmp/deepsea-tmp.tar HEAD
+	HASH=`git rev-parse --short HEAD`; \
+	OFFSET=`git rev-list \`git log --pretty=oneline -- version.txt | head -1 | cut -d ' ' -f 1\`..HEAD | wc -l`; \
+	cd .tmp; tar -xf deepsea-tmp.tar; cd deepsea-tmp; \
+	VERSION=`cat version.txt`; F_VERSION="$$VERSION+git.$$OFFSET.$$HASH"; \
```
Comment (smithfarm) [inline Makefile]: In ceph the RPM version number format is `x.y.z-offset-sha1` (e.g. 13.0.0-3504-ga292d52fe7) and for this I have verified that RPM considers e.g. the following series to be ascending:

1. `13.0.0-9-g$SHA1`
2. `13.0.0-10-g$SHA1`

However, I'm not sure if RPM will be so savvy with this series:

1. `13.0.0+git.9.$SHA1`
2. `13.0.0+git.10.$SHA1`

That will need testing before this PR is merged?

**Comment by rjfd** [comment]
@LenzGr @ricardoasmarques I just added to this PR a commit with a runner that returns the version info.

File: Makefile (hunk 2 above)
Comment (LenzGr) [reply Makefile]: I personally am not a big fan of adding git revision hashes to version numbers, as they are random and don't really indicate any ordering. A timestamp is much more readable IMHO - the corresponding git revision hash could still be stored in the version.txt file, of course.

**Comment by rjfd** [comment]
This PR is also related with the openATTIC tracker issue: https://tracker.openattic.org/browse/OP-2543

File: srv/modules/runners/deepsea.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+import os
+import re
+
+
+def version():
+    """
+    Returns the DeepSea version info currently installed
+    """
+    version_path = '/usr/share/doc/packages/deepsea/version.txt'
```
Comment (LenzGr) [inline srv/modules/runners/deepsea.py]: A hardcoded path sounds risky to me.

On non SUSE-platforms, this would be `/usr/share/doc/deepsea/version.txt`, for example (if RPM's `%docdir` macros is utilized). Wouldn't it be better to keep it somewhat "closer" to the DeepSea code?

Comment (rjfd) [reply srv/modules/runners/deepsea.py]: Good catch. I would be in favor of storing it in the pillar. 
@swiftgist what do you think?

File: srv/modules/runners/deepsea.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+This runner implements helper functions for DeepSea in general
+"""
+
+from __future__ import absolute_import
+
+import os
+import re
+
+
+def version():
```
Comment (ricardoasmarques) [inline srv/modules/runners/deepsea.py]: Should be `def version(**kwargs):`.

Without `**kwargs` this runner cannot be called from the **salt-api**.

File: Makefile (hunk 2 above)
Comment (smithfarm) [reply Makefile]: @LenzGr Having the git hash/SHA1 in the version number is useful for developers because it enables easy cross-linking between "code in RPM" and "code in git". However, it's useless for ensuring that the RPM version number is always increasing. That's why the timestamp is sometimes used, but git offset is another option.

It's interesting to note here that the OBS replaces whatever is in `Release: ...` in the spec file with its own, always ascending, number (the OBS revision number). So, as long as we populate the `Version: ...` line in the spec file with *just* the major version number (e.g. `0.8.0`) and put the other components (offset, sha1, timestamp, etc.) in `Release: ...`, it doesn't matter because OBS ensures that each successive RPM version number will be greater than the one before.

Comment (smithfarm) [reply Makefile]: For an example of how this works, see https://build.opensuse.org/package/show/filesystems:ceph:mimic/ceph -- in the spec file, the "Release:" line says "3535.g8b10d08c04" (git offset and hash), but the built RPMs look like this: `ceph-13.0.0-194.1.src.rpm` because, during the build, OBS does this:

```
[  113s] -----------------------------------------------------------------
[  113s] I have the following modifications for ceph.spec:
[  113s] 71c71
[  113s] < Release:	3535.g8b10d08c04%{?dist}
[  113s] ---
[  113s] > Release:	194.1
[  113s] -----------------------------------------------------------------
```

File: srv/modules/runners/deepsea.py (hunk 4 above)
Comment (rjfd) [reply srv/modules/runners/deepsea.py]: done!

File: Makefile (hunk 2 above)
Comment (jan--f) [reply Makefile]: > However, I'm not sure if RPM will be so savvy with this series:
>
>    1. 13.0.0+git.9.$SHA1
>    2. 13.0.0+git.10.$SHA1

This is the deepsea version string in the package name. Has worked fine so far. I'll double check.

**Comment by rjfd** [comment]
@jan--f I'm also preparing a change on our OBS build scripts to work with this. Last friday I sent an email to deepsea-users describing the whole new build process but it seems that that email was never delivered. Today I'll resend that email where I describe the new OBS build process.

**Comment by jan--f** [comment]
Ok I might be missing something but all this strikes me as very complex and cumbersome. We should discuss this in more details at the standup.
Iiuc this change aims to install the version.txt file. To achieve this we maintain a version.txt in the repo and change the package building process considerably. I do have issues with that:

- So far a new git tag is enough to create a new version. Now  we need to create a tag and change the version in the file.
- This forces us to maintain a build artefact outside of the repo. Granted we do this already but the _service file is declarative, checkin.sh is code. How will we track and fix bugs?

I think we can achieve the same thing with the current build process by small additions to the spec file and Makefile.

Also I understand that the aim is to have the deepsea runner return the version. Why do we not simply put the version string into the runner, no extra files needed. Ideally all our python files would have this version build in tbh, but that's another topic.
So am I missing something crucial? I'm happy to propose a PR of what I'd consider a better approach, unless I missing the point of course.

**Comment by rjfd** [comment]
@jan--f in this new approach the git tag is not needed, the only source of the version number is the `version.txt`. Also, the usage of the git tag to track the version number is more error prone (requires more discipline) and if someone pushes a new tag it will even change the git offset when generating the version number from `git describe`.

Another important issue is the support for the packaging in other distros. We should make the whole "release process" the more distro independent as possible. So, if we start doing a lot of magic steps inside the spec file, we will become more dependent on RPM based distros for instance.

> I'm happy to propose a PR of what I'd consider a better approach, unless I missing the point of course.

Alternative approaches are always welcome :-)

File: deepsea.spec.in [hunk 5]
```diff
@@ -22,7 +22,7 @@
 # See also http://en.opensuse.org/openSUSE:Shared_library_packaging_policy
 
 Name:           deepsea
-Version:        0.8
+Version:        @VERSION@
 Release:        0
```
Comment (smithfarm) [inline deepsea.spec.in]: You might want to check if the (rather complicated) version string you intend to use here is compatible with both DEB and RPM. As I mentioned before (i.e. sorry for repeating myself but it might be important), upstream Ceph chose to split the components by putting only the version number in "Version:" and the rest in "Release:".

I do think this approach might end up being more "universal" but I don't (yet) have any evidence or hard arguments to prove that.

Comment (rjfd) [reply deepsea.spec.in]: @smithfarm thanks for the reminder. I know that in opensuse that version string is currently working good as it's the format already used for DeepSea for several months now.
I'll need to check if that works in other distros, including debian based one.

Comment (smithfarm) [reply deepsea.spec.in]: After reading https://www.debian.org/doc/debian-policy/#s-f-version I *think* the full version.txt version number will work as a Debian "upstream_version" number, i.e. without use of "debian_revision". So my comment above can be safely disregarded.

Comment (smithfarm) [reply deepsea.spec.in]: @rjfd https://www.debian.org/doc/debian-policy/#s-f-version has a very nice explanation of how Debian does the version comparison. I think you'll be fine.

**Comment by jan--f** [comment]
Ok I took a stab at this with the existing build process in mind. https://github.com/SUSE/DeepSea/compare/jan-version-packaging?expand=1

**Comment by rjfd** [comment]
@jan--f 
> https://github.com/SUSE/DeepSea/compare/jan-version-packaging?expand=1

The main difference between our approaches is in how the version number is managed. In my approach the version number is stored in a `version.txt` file, and it's required a new commit to change it. And in your approach the version is stored as a git tag and excludes the existence of a `version.txt` file with the version number.

In your approach you are also not updating the version number in the spec file, because you are relying on the OBS service to update it. I think that's a bad thing because you'll have two distinct procedures to generate the version string (the Makefile, and the OBS service), which now are similar but we never know when they may diverge.

In my approach, the tarball has everything ready to build with the correct version numbers, including the distro specific build files, and therefore it doesn't depend on OBS to do the work. That's why I dropped the usage of the OBS service in favor of a bash script that can easily be used to generate the necessary files for packaging DeepSea in OBS.
See: https://build.opensuse.org/package/view_file/home:rjdias:branches:filesystems:ceph/deepsea/checkin.sh?expand=1

> I think. E.g. make install currently leaves an unstaged change in setup.py. There are multiple ways of solving this...pending discussion.

I think that any change to the source files should be done only in the files inside the tarball, and that's is what I'm already doing in my approach. Now that I see in more detail the code of your approach, it does not work because `git archive` will only tar files that are committed, therefore the changes you made with `sed` will not be part of the tarball.

The bottom line of my analysis is: you referred to my approach as
> "all this strikes me as very complex and cumbersome."

and although I don't quite understand if you are either saying that it's the code that is complex or the packaging that it's complex, in order for your approach to work will also need to become a bit more "complex".

Therefore, I'm of the opinion that what we need to decide among the DeepSea contributors is that if we want to store the version number in a file inside the tarball or store the version as a `git tag`. 
@swiftgist @smithfarm @jschmid1 @LenzGr @kmroz ...

**Comment by jan--f** [comment]
Ok I added the version replacement to the tarball make target.

I'm not against the version.txt per se. If there is benefit, lets add it.

I think its not a good idea to make version.txt the authoritative source of the version.

The branch I provided is just a POC. As I said there are warts that need to be removed (and can be easily removed). I think the benefit is that everything needed for packaging is contained in the Makefile.

**Comment by smithfarm** [comment]
@jan--f IIRC DeepSea's present system for setting the version number and generating the tarball was inspired by the one used by SES. This, in turn, relied on upstream's practice of adding a tag (e.g. "v12.2.1") whenever they increment the version number.

As long as these "official" version tags are the only tags ever added to the repository, this system works well. The problems start when you want to add other tags (in SES's case, Milestone tags for example).

The command "git describe" returns a string containing three components:

1. the most recent tag in the current branch
2. the number of commits in the branch between that tag and HEAD
3. the SHA1 of HEAD

This string is parsed to extract the components and, as I said above, as long as only the version tags are there, the system works fine.

However, it's easy to see how the system can break. Consider the following "real-life" scenario:

1. the project lead decides it's time to release Version 6.7.8
2. a "v6.7.8" tag is added to the repo
3. tarball is generated, packages built, etc. - all good
4. time passes, commits are added to master
5. more tarballs are generated, packages built - still good
6. somebody (this can be anybody, for any reason) adds a tag "foo" to the branch
7. the next time you try to build packages -> KABOOM

If this isn't immediately clear, try the following commands in the master branch of ceph/ceph.git:

```
$ git describe --tags
v13.0.0-3542-g6863b4914e
$ git tag Milestone12 HEAD^^^
$ git describe --tags
Milestone12-6-g6863b4914e
```

Obviously, we do not want "Milestone12", or any "foo", as the version number.

In Ceph we are mitigating this by using `git describe --match 'v*' | sed 's/^v//'` [1] - this makes the problem much less severe since only annotated tags starting with "v" are considered. Still, this whole system strikes me as unnecessarily fragile and if I were designing the workflow I would not rely on git tags for the version number at all. (In my case, the matter is moot because I have to follow upstream, but DeepSea has the luxury of being "its own upstream"!)

That's why I'm in favor of @rjfd 's proposal.

[1] https://github.com/SUSE/ceph/blob/ses6/make-dist#L9

**Comment by jan--f** [comment]
@smithfarm all true and correct. I just don't see how manually maintaining a version.txt improves this.

**Comment by smithfarm** [comment]
@jan--f My comment was my attempt to argue for maintaining the version number in a file inside the tarball, instead of as a git tag. Quoting @rjfd 's comment:

> Therefore, I'm of the opinion that what we need to decide among the DeepSea contributors is that if we want to store the version number in a file inside the tarball or store the version as a git tag.

As @LenzGr already said, the version number needs to be stored in one (and only one) place. I think we can all agree on this, at least. (?)

If, as per my argumentation above, the git tag is not appropriate for maintaining the version number, then why not use version.txt for the same?

You ask why "manually" maintaining version.txt helps - I guess the emphasis here was on the word "manually"? Changing in the version number is an inherently manual process since it relies on human judgment. For example, if the version number is 6.7.8 and we want to increment it, there are at least three possibilities: 6.7.9, 6.8, and 7.0.

File: Makefile [hunk 6]
```diff
@@ -585,8 +587,18 @@ rpm: tarball test
 
 # Removing test dependency until resolved
 tarball:
-	VERSION=`awk '/^Version/ {print $$2}' deepsea.spec`; \
-	git archive --prefix deepsea-$$VERSION/ -o ~/rpmbuild/SOURCES/deepsea-$$VERSION.tar.bz2 HEAD
+	mkdir -p .tmp
+	git archive --prefix deepsea-tmp/ -o .tmp/deepsea-tmp.tar HEAD
+	HASH=`git rev-parse --short HEAD`; \
+	OFFSET=`git rev-list \`git log --pretty=oneline -- version.txt | head -1 | cut -d ' ' -f 1\`..HEAD | wc -l`; \
```
Comment (smithfarm) [inline Makefile]: @jan--f This line is key, I think. Instead of using `git describe`, @rjfd is calculating the offset as the number of commits since the last commit that changed version.txt.

**Comment by swiftgist** [comment]
Okay, I did some discussing and reading yesterday/today.  The short version seems to be that I should be using signed, annotated git tags.  That's why they're there.  The reason to use 'git tag -> version' rather than 'version -> git tag' is git supports gpg.  So, that's a bit more natural.

I do prefer Jan's approach in "hardcoding" the version numbers.  Reminds me of what we do with binaries (if we had them), but it also means less moving parts at runtime.  To address a couple of concerns, I'll reference his branch.

Change the 'git describe --tags' to 'git describe'.  This allows anyone to still use lightweight (i.e. normal) tags without causing issues for building packages. 

Another addition is to save the output of the modified text to "version.txt" for the source tarball.  It won't be used beyond that and it matches the tarball filename, etc.

**Comment by smithfarm** [comment]
> Change the 'git describe --tags' to 'git describe'. This allows anyone to still use lightweight (i.e. normal) tags without causing issues for building packages.

It might be interesting to look which version of git added the `--tags` behavior. It might be a fairly recent thing.

**Comment by rjfd** [comment]
@smithfarm @swiftgist @jan--f please check this other branch where I implemented a solution based on "git annotated tags" and where version strings are inlined in the pythons files:

https://github.com/SUSE/DeepSea/compare/master...rjfd:wip-version-number2?expand=1

If you're good with this new implementation I'll push the commits to this PR.

**Comment by LenzGr** [comment]
@rjfd looks good to me!

**Comment by rjfd** [comment]
test this please

**Comment by rjfd** [comment]
please test this

**Comment by kmroz** [comment]
Another sound approach.  I do like the tighter coupling of the version with DeepSea in #854 (and as I mentioned there, it's cool that both you and @jan--f are on it, given you were both working to solve the same problem).

Whatever the powers decide...

**Comment by ricardoasmarques** [comment]
Shouldn't this PR be closed, since it was replaced by https://github.com/SUSE/DeepSea/pull/854 ?
